linera-execution/src/evm/revm.rs

```diff
@@ -1001,16 +773,64 @@ impl<Runtime: ContractRuntime> CallInterceptorContract<Runtime> {
             // The EVM smart contract is being called
             return Ok(None);
         }
+        // Handling the balances.
+        if let CallValue::Transfer(value) = inputs.value {
+            let source: AccountOwner = inputs.caller.into();
+            let owner: AccountOwner = inputs.bytecode_address.into();
+            if value != U256::ZERO {
+                // In Linera, only non-zero transfers matter
+                let mut runtime = context
+                    .db()
+                    .0
+                    .runtime
+                    .lock()
+                    .expect("The lock should be possible");
+                let amount = Amount::try_from(value).map_err(EvmExecutionError::from)?;
+                let chain_id = runtime.chain_id()?;
+                let destination = Account { chain_id, owner };
+                runtime.transfer(source, destination, amount)?;
+            }
+            // In Ethereum, all transfers matter
+            if let Some(i) =
+                context
+                    .journal()
+                    .transfer(inputs.caller, inputs.target_address, value)?
+            {
+                let error = format!("{i:?}");
+                let error = EvmExecutionError::TransactError(error);
+                return Err(ExecutionError::EvmError(error));
+            }
+        }
         // Other smart contracts calls are handled by the runtime
         let target = address_to_user_application_id(inputs.target_address);
-        let argument = get_call_argument(context, inputs);
-        let authenticated = true;
-        let result = {
-            let mut runtime = self.db.runtime.lock().expect("The lock should be possible");
-            runtime.try_call_application(authenticated, target, argument)?
+        let (argument, n_input) = get_call_contract_argument(context, inputs)?;
+        let result = if n_input > 0 {
+            // The input is non-trivial, we assume that we are calling a contract.
+            //
+            // The correct behavior is the following:
+            // * If a contract exists and input is empty, then the fallback function is called
+            //   if existing. Otherwise failure.
+            // * If not, then it is a user account and no execution occurs.
+            //
+            // So, the correct way is instead to test the existence of the application
+            // given the application_id. So, we need following function in BaseRuntime:
+            //  fn is_existing_application(&mut self, application_id: ApplicationId) -> Result<bool, ExecutionError>;
```

Comment:
Ideally we'd also handle the case where the application has not _yet_ been used on this chain, but will be. So we'd transfer the balance to the account as if it were a user account, but then, if the application is later used, it should have the correct balance. Or is that already the case?

Reply:
Yes, I can handle that case in a different way. Will see later.